### ISSUE TITLE
Xava randomly quits working

### ISSUE THREAD

**Issue by klandrith** [nikp123/xava#26]
Randomly (usually after quite a while of no audio activity) xava quits working. I have to kill it and restart it in order for it to work properly. If music is going, it will work for hours on end without problem.
Here is my setup.

Manjaro KDE (all current)
-Plasma 5.19.5
-Qt 5.15.1
-Kernel 5.8.11-1

**Comment by nikp123** [comment]
QT version is unrelated, but would you mind telling me what audio backend do you use?

**Comment by nikp123** [comment]
It narrows down the bug search by quite a lot.

**Comment by klandrith** [comment]
PulseAudio. Thanks for the quick reply!

**Comment by nikp123** [comment]
For how long do you have to wait in order to find this issue

**Comment by klandrith** [comment]
I have this is the audio section of my xava config file.

method = pulse
source = auto

**Comment by nikp123** [comment]
yes, but how long did you wait for?

**Comment by klandrith** [comment]
Well I mean I'm unsure of that because I never really pay attention. But I would say it happens after an hour or more of no audio activity. I recently setup a KVM Windows 10 virtual machine, and I have been switching back and forth quite a bit between it and my Linux desktop. Usually I notice the bug after I've been on the VM for awhile and switched back over the the my main desktop. Audio output never has any problem between switching between the guest and host.

**Comment by nikp123** [comment]
What audio setup are you using in your VM?

I think I know what's going on here: PulseAudio sink changes due to the VM and XAVA can't handle that.

**Comment by nikp123** [comment]
tl;dr: QEMU likes to mess with PulseAudio

**Comment by klandrith** [comment]
Makes sense, I was suspecting as much. I'm using PulseAudio for the VM. I wish I could just pass in my whole audio device, but it is non-resettable, so I can't do that. I will have to explore other audio options for the VM. Doesn't seem to be a problem on xava's end then, really.

BTW,
Great piece of software. It is absolutely awesome. But, is there any way to improve the graphical performance of xava though? It uses a lot of GPU resources (I only have a Nvidia GT1030 on my host)? I know that question is unrelated...

**Comment by nikp123** [comment]
@klandrith It is (to some extent) XAVA's fault, but the behaviour is basically non-standard, if you get what I'm saying ;)

Just I'm not sure if applications can detect a dead source.

Also can you share your QEMU audio params, it could help a lot.

**Comment by nikp123** [comment]
> Makes sense, I was suspecting as much. I'm using PulseAudio for the VM. I wish I could just pass in my whole audio device, but it is non-resettable, so I can't do that. I will have to explore other audio options for the VM. Doesn't seem to be a problem on xava's end then, really.
> 
> BTW,
> Great piece of software. It is absolutely awesome. But, is there any way to improve the graphical performance of xava though? It uses a lot of GPU resources (I only have a Nvidia GT1030 on my host)? I know that question is unrelated...

Try disabling OpenGL. I've already had like 3 issues, all NVIDIA related. It speaks for itself. You'll lose shadows, unfortunately.

**Comment by klandrith** [comment]
I gotcha. 

This is all that I have in my audio section of my VM libvirt xml file when it comes to sound...

<sound model="ich9">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x1b" function="0x0"/>
    </sound>

I just added an emulated sound device via the libvirt GUI. I am exploring other configurations right now. I *could* use scream for audio, but I'm not sure if that would suffer the same fate or not?

I'll try disabling OpenGL. I don't care for the shadows anyways as my desktop background is very dark and I can't see them anyways. lol.

**Comment by nikp123** [comment]
Scream work well if you don't pin the CPU, then you get artifacting.

**Comment by klandrith** [comment]
Bummer, I am pinning my CPU cores...ugh.

**Comment by klandrith** [comment]
"<sound model="ich9">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x1b" function="0x0"/>
    </sound>"

Forgot to quote that...That's all I have in my XML file for audio.

**Comment by nikp123** [comment]
not the device in qemu, but how QEMU plays the audio itself?

To make it easier for you just share the entire thing, if you don't mind.

**Comment by klandrith** [comment]
`<domain xmlns:qemu="http://libvirt.org/schemas/domain/qemu/1.0" type="kvm">
  <name>win10</name>
  <uuid>9a735831-13fd-4cfd-b50f-f86e1d06e368</uuid>
  <metadata>
    <libosinfo:libosinfo xmlns:libosinfo="http://libosinfo.org/xmlns/libvirt/domain/1.0">
      <libosinfo:os id="http://microsoft.com/win/10"/>
    </libosinfo:libosinfo>
  </metadata>
  <memory unit="KiB">33554432</memory>
  <currentMemory unit="KiB">33554432</currentMemory>
  <memoryBacking>
    <hugepages/>
    <nosharepages/>
    <locked/>
  </memoryBacking>
  <vcpu placement="static">20</vcpu>
  <cputune>
    <vcpupin vcpu="0" cpuset="0"/>
    <vcpupin vcpu="1" cpuset="16"/>
    <vcpupin vcpu="2" cpuset="1"/>
    <vcpupin vcpu="3" cpuset="17"/>
    <vcpupin vcpu="4" cpuset="2"/>
    <vcpupin vcpu="5" cpuset="18"/>
    <vcpupin vcpu="6" cpuset="3"/>
    <vcpupin vcpu="7" cpuset="19"/>
    <vcpupin vcpu="8" cpuset="4"/>
    <vcpupin vcpu="9" cpuset="20"/>
    <vcpupin vcpu="10" cpuset="5"/>
    <vcpupin vcpu="11" cpuset="21"/>
    <vcpupin vcpu="12" cpuset="6"/>
    <vcpupin vcpu="13" cpuset="22"/>
    <vcpupin vcpu="14" cpuset="7"/>
    <vcpupin vcpu="15" cpuset="23"/>
    <vcpupin vcpu="16" cpuset="8"/>
    <vcpupin vcpu="17" cpuset="24"/>
    <vcpupin vcpu="18" cpuset="9"/>
    <vcpupin vcpu="19" cpuset="25"/>
    <emulatorpin cpuset="10-11,26-27"/>
  </cputune>
  <os>
    <type arch="x86_64" machine="pc-q35-5.1">hvm</type>
    <loader readonly="yes" type="pflash">/usr/share/edk2-ovmf/x64/OVMF_CODE.fd</loader>
    <nvram>/var/lib/libvirt/qemu/nvram/win10_VARS.fd</nvram>
    <bootmenu enable="yes"/>
  </os>
  <features>
    <acpi/>
    <apic eoi="on"/>
    <hap state="on"/>
    <hyperv>
      <relaxed state="on"/>
      <vapic state="on"/>
      <spinlocks state="on" retries="8191"/>
      <vpindex state="on"/>
      <vendor_id state="on" value="AuthenticAMD"/>
      <frequencies state="on"/>
    </hyperv>
    <kvm>
      <hidden state="on"/>
      <hint-dedicated state="on"/>
    </kvm>
    <ioapic driver="kvm"/>
  </features>
  <cpu mode="host-passthrough" check="none" migratable="on">
    <topology sockets="1" dies="1" cores="10" threads="2"/>
    <cache mode="passthrough"/>
    <feature policy="require" name="invtsc"/>
    <feature policy="require" name="topoext"/>
    <feature policy="disable" name="monitor"/>
  </cpu>
  <clock offset="utc">
    <timer name="rtc" present="no" tickpolicy="catchup"/>
    <timer name="pit" present="no" tickpolicy="discard"/>
    <timer name="hpet" present="no"/>
    <timer name="tsc" present="yes" mode="native"/>
    <timer name="hypervclock" present="yes"/>
  </clock>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <pm>
    <suspend-to-mem enabled="no"/>
    <suspend-to-disk enabled="no"/>
  </pm>
  <devices>
    <emulator>/usr/bin/qemu-system-x86_64</emulator>
    <disk type="file" device="cdrom">
      <driver name="qemu" type="raw"/>
      <source file="/var/lib/libvirt/images/virtio-win-0.1.185.iso"/>
      <target dev="sdb" bus="sata"/>
      <readonly/>
      <boot order="2"/>
      <address type="drive" controller="0" bus="0" target="0" unit="1"/>
    </disk>
    <controller type="usb" index="0" model="qemu-xhci" ports="15">
      <address type="pci" domain="0x0000" bus="0x02" slot="0x00" function="0x0"/>
    </controller>
    <controller type="sata" index="0">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x1f" function="0x2"/>
    </controller>
    <controller type="pci" index="0" model="pcie-root"/>
    <controller type="pci" index="1" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="1" port="0x8"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x0" multifunction="on"/>
    </controller>
    <controller type="pci" index="2" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="2" port="0x9"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x1"/>
    </controller>
    <controller type="pci" index="3" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="3" port="0xa"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x2"/>
    </controller>
    <controller type="pci" index="4" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="4" port="0xb"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x3"/>
    </controller>
    <controller type="pci" index="5" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="5" port="0xc"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x4"/>
    </controller>
    <controller type="pci" index="6" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="6" port="0xd"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x5"/>
    </controller>
    <controller type="pci" index="7" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="7" port="0xe"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x6"/>
    </controller>
    <controller type="pci" index="8" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="8" port="0xf"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x7"/>
    </controller>
    <controller type="pci" index="9" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="9" port="0x10"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x0" multifunction="on"/>
    </controller>
    <controller type="pci" index="10" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="10" port="0x11"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x1"/>
    </controller>
    <controller type="pci" index="11" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="11" port="0x12"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x2"/>
    </controller>
    <controller type="pci" index="12" model="pcie-to-pci-bridge">
      <model name="pcie-pci-bridge"/>
      <address type="pci" domain="0x0000" bus="0x0b" slot="0x00" function="0x0"/>
    </controller>
    <controller type="pci" index="13" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="13" port="0x13"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x3"/>
    </controller>
    <controller type="pci" index="14" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="14" port="0x14"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x4"/>
    </controller>
    <input type="keyboard" bus="virtio">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x0f" function="0x0"/>
    </input>
    <input type="mouse" bus="ps2"/>
    <input type="keyboard" bus="ps2"/>
    <sound model="ich9">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x1b" function="0x0"/>
    </sound>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x0d" slot="0x00" function="0x0"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x03" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x0d" slot="0x00" function="0x1"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x04" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x0d" slot="0x00" function="0x2"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x05" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x0d" slot="0x00" function="0x3"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x06" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x04" slot="0x00" function="0x0"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x07" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x05" slot="0x00" function="0x0"/>
      </source>
      <boot order="1"/>
      <address type="pci" domain="0x0000" bus="0x08" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x0f" slot="0x00" function="0x3"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x0a" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="usb" managed="yes">
      <source>
        <vendor id="0x8087"/>
        <product id="0x0029"/>
      </source>
      <address type="usb" bus="0" port="1"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x09" slot="0x00" function="0x0"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x01" slot="0x00" function="0x0"/>
    </hostdev>
    <hostdev mode="subsystem" type="pci" managed="yes">
      <source>
        <address domain="0x0000" bus="0x0c" slot="0x00" function="0x0"/>
      </source>
      <address type="pci" domain="0x0000" bus="0x09" slot="0x00" function="0x0"/>
    </hostdev>
    <memballoon model="none"/>
    <panic model="hyperv"/>
  </devices>
  <qemu:commandline>
    <qemu:arg value="-object"/>
    <qemu:arg value="input-linux,id=kbd1,evdev=/dev/input/by-id/usb-Logitech_G512_Carbon_Tactile_198238753531-event-kbd,repeat=on"/>
  </qemu:commandline>
</domain>
`
For audio on the guest I am using the audio on my 2080ti via displayport to my monitor. I haven't tried to output audio via my audio jacks, so I haven't really bothered on that end. I'm not sure my setup even works on that front. I am new to this KVM stuff, as I just got the VM up and running last week.

**Comment by klandrith** [comment]
Sorry that's a mess, not sure how to format it to keep the indents and such..

**Comment by nikp123** [comment]
yeah, it doesn't reveal much.....

**Comment by klandrith** [comment]
What part of the qemu xml exactly are you looking for? Sorry for my lack of knowledge about this stuff. As I said, just got it going...My head is still spinning from the 100 or so guides I read last week, lol.

**Comment by nikp123** [comment]
don't bother yourself with the details, it's mostly dev jargon anyway

the thing is I just need to be sure what the culprit is

**Comment by nikp123** [comment]
For example, I booted up my KVM machine, and it works just fine now

**Comment by klandrith** [comment]
Yah, I gotcha. That was a classic line in my CS programming and algorithms 1 class. lol "ran on my machine..."

**Comment by nikp123** [comment]
also can you try the following dumb hack (if it works):

change the starting ``<domain>`` tag into ``<domain type='kvm' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>``
before saving, also add:
```
  <qemu:commandline>
    <qemu:arg value='-audiodev'/>
    <qemu:arg value='pa,id=pa1,server=/run/user/1000/pulse/native'/>
 </qemu:commandline>
```
BEFORE the final ``</domain>`` tag.

**Comment by nikp123** [comment]
this is just to replicate my audio setup, to see if there's something about QEMU that messes with it

**Comment by nikp123** [comment]
oh wait nevermind
you don't have to change the tag, it's already done for you

**Comment by nikp123** [comment]
but just add the second part

**Comment by klandrith** [comment]
Yah, I gotcha. I tried that and I get an error in libvirt that it was unabled to connect to pulseaudio. 

Error starting domain: internal error: qemu unexpectedly closed the monitor: pulseaudio: pa_context_connect() failed
pulseaudio: Reason: Connection refused
pulseaudio: Failed to initialize PA contextaudio: Could not init `pa' audio driver

**Comment by klandrith** [comment]
Nevermind, figured it out, need to change my user in qemu config file. I was(am) running a diff user to access my keyboard via evdev.

**Comment by nikp123** [comment]
boom, you just found the problem yourself.

That means that pulseaudio runs as root then, and xava gets rejected access to audio (for obvious reasons)

**Comment by nikp123** [comment]
should error out then, should it?

**Comment by nikp123** [comment]
:thonk:

**Comment by nikp123** [comment]
But there's a workaround for this, it's just not device agnostic... (developer speak for it doesn't work when you plug in an another device) Use alsa instead. It's complicated to set up, but it avoids problems like these.

**Comment by klandrith** [comment]
Yah, I think we got it nailed down. Thanks for all the help! Now I'm just trying to get hda to work in libvirt. Libvirt is complaining that it can't find 'hda'. 

Use alsa for xava?

**Comment by nikp123** [comment]
Yes, but if you really sure about using it. It's difficult to set up.

**Comment by nikp123** [comment]
Or instead, you can write a udev file that grants access to your keyboard using a certain group instead of a user.

**Comment by klandrith** [comment]
I got it working now. Everything is going fine. Just started up the VM, switched monitor inputs to VM, audio cutoff, switched back, audio resumed and xava was still working. So it seems to be resolved. Thanks again for all the help, even though this didn't turn out to be an issue with xava!

**Comment by klandrith** [comment]
If I still have an issue with xava cutting out again, I'll post back to this issue thread, but I think we've got it all figured out.

**Comment by nikp123** [comment]
Yeah, I'm glad there are people testing out my code. Gives me motivation to work on the project.

Thanks for you input and happy visualizing  (don't worry, you can always reopen this issue later on)

**Comment by nikp123** [comment]
[99-qemu-hw-users.rules.zip](https://github.com/nikp123/xava/files/5324323/99-qemu-hw-users.rules.zip)
also here's that udev file in case you need it @klandrith
